akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/TestKitStorage.scala

```diff
@@ -47,15 +51,24 @@ sealed trait InMemStorage[K, R] extends InternalReprSupport[R] {
           Some(value.drop(fromInclusive).take(maxNum))
         else None)
 
+  def removeFirst(key: K): Unit = Option(eventsLog.get(key)).foreach(_.poll())
+
+  def first(key: K): Option[R] = Option(eventsLog.get(key)).flatMap { item =>
+    Try(item.element()).toOption.map(toRepr)
+  }
+
   def findOneByIndex(key: K, index: Int): Option[R] =
     Option(eventsMap.get(key))
       .flatMap {
         case (_, value) => if (value.size > index) Some(value(index)) else None
       }
       .map(toRepr)
 
-  def add(key: K, p: R): Unit =
+  def add(key: K, p: R): Unit = {
+    eventsLog.putIfAbsent(key, new util.LinkedList[InternalRepr]())
+    eventsLog.get(key).add(toInternal(p))
```

Comment:
The design of using concurrent data structures for the InMemStorage is not great and that is highlighted by that it's too difficult to make changes here. Since this is "only" TestKit I think we should change to using `synchronized` access and ordinary collections. Let me know if I should take over this PR and do the bigger concurrency refactoring?

Reply:
How about this: we bring this PR over the finish line and we can do a follow up refactoring PR to separate scopes? In any case I am happy make the changes related to the `synchronized` approach.